### ISSUE TITLE
FEATURE REQUEST - GRAD CAM FOR YOLO

### ISSUE THREAD

**Issue by holger-prause** [AlexeyAB/darknet#7306]
Hello,

For debugging a model its a very userfull to show a heatmap of the classactivations.
This is described in paper https://arxiv.org/pdf/1610.02391v1.pdf
* Implementation in torch: https://github.com/ramprs/grad-cam
* Implementation in pytorch https://github.com/jacobgil/pytorch-grad-cam 

This would be very usefull imho.

**Comment by AlexeyAB** [comment]
https://github.com/AlexeyAB/darknet/issues/5117
Set in your cfg-file and run training:
```
[net] 
adversarial_lr=0.05 
attention=1
```

**Comment by holger-prause** [comment]
Thank you very much :-)
Do i need to specify **-show_imgs** and will this produce a heatmap?

**Comment by holger-prause** [comment]
Oh i am stupid - i just have to read the linked issue #5517 - its all there!
Thank you very much again - yolo is awesome!

**Comment by holger-prause** [comment]
@AlexeyAB 
My goal is to visualize the classactivation for a certain class - even the classactivationmap for all classes would be fine.
Before testing my own model - i want to try out on sample image dog.jpg and the coco weights (yolov4.weights) and try to get 
the attention visualization.

I created a small dataset containing the dog.jpg sample image. ** I 
dont created any annotation for this sample**.

After this i set the mentioned config parameter into the net section
```
    [net] 
    adversarial_lr=0.05 
    attention=1
```

After this is started training with -
**darknet_no_gpu.exe detector train C:\development\datasets\gradcamtest\train.data C:\development\datasets\gradcamtest\yolov4.cfg C:\development\datasets\gradcamtest\yolov4.weights -show_imgs**

After this - i see a bunch auf augemented images - but none of them contains any kind of heatmap or attention visualization.
What am i doing wrong (should i pick yolov4.conv.137?, do i need to picked annotated images?) ?

Thank you very much for your patience - i did some trainings before - for me this is not really straighforward and any help is highly welcome.

**Comment by AlexeyAB** [comment]
* Dont use -show_imgs.
* Set burn_in=0 https://github.com/AlexeyAB/darknet/blob/64efa721ede91cd8ccc18257f98eeba43b73a6af/cfg/yolov4.cfg#L18

**Comment by holger-prause** [comment]
@AlexeyAB 
Ok i did the suggest changes and removed -show_imgs and added the burn in - the console output is
`    Saving weights to C:\development\datasets\gradcamtest\backup/yolov4_final.weights`

But i still dont see any attention image generated(i checked dataset backup and darknet folder). Note that i use the no gpu version. Am i missing something very obvious - do i need to annotate my sample image?

**Comment by AlexeyAB** [comment]
It supports on GPU. It isn't implemented for CPU.

**Comment by holger-prause** [comment]
@AlexeyAB 
Ah ok - i was assuming something like this - fair enough, i got a new pc on work and it has no cuda gpu...
I train in cloud.

Can i please summary(my understanding) what i need to do to get the attention image:
**First of all: Make sure you use gpu (cuda) variant of yolo framework**
* Step 1 - Create a yolo dataset ready to train with the sample image(s) on which you want to do the prediction
* Step 2 - Adopt you model config to include (as mentioned above)
```
    [net] 
    adversarial_lr=0.05 
    attention=1 
```  

AND also 
```
    burn_in=0 
```
* Step 3 - Run training (dont use -show_img.!)

Now i can only guess:
* Step 4?
Use generated weights file and adopted config to do a prediction (inference)

Is this correct? 
Thank you again very much for your time and patience.

**Comment by AlexeyAB** [comment]
1. Train the model or use already trained cfg/weights files
2. Change in cfg `[net]    adversarial_lr=0.05     attention=1  burn_in=0`
3. Run training on GPU with these cfg/weights files  (you can tune adversarial_lr=0.05)
    * just additional not: it will generate new weights in the /backup/ directory by using SAT (self-adversarial trainin), it can be better if  model(yolov4) has higher capacity than dataset(ua-detrac), and can be worse for model(yolov4) has lower capacity than dataset(mscoco)

**Comment by holger-prause** [comment]
Ok thank you very much - but its still unclear when the attention images are generated - during training process or during inference with the custom created weights. But this is can find out by triyng out myself or looking at the source code.
Thank you very much - have a nice weekend.

**Comment by holger-prause** [comment]
OMG its finally working - i tried out on another computer and i had to use -clear.
Make sure it is training for at least some batches. Next thing would be to get attention for only one class - but i already see different colors for different objects - this ill find out myself.

**Comment by bjajoh** [comment]
@holger-prause I'm not sure if I'm understanding it correctly, but is there a working Grad Cam implementation to give me the heat map?

**Comment by jojo0513** [comment]
> 1. Train the model or use already trained cfg/weights files
> 2. Change in cfg `[net]    adversarial_lr=0.05     attention=1  burn_in=0`
> 3. Run training on GPU with these cfg/weights files  (you can tune adversarial_lr=0.05) and with flag `-clear` at the end of training command
>    
>    * just an additional note: it will generate new weights in the /backup/ directory by using SAT (self-adversarial trainin), it can be better if  model(yolov4) has higher capacity than dataset(ua-detrac), and can be worse for model(yolov4) has lower capacity than dataset(mscoco)

Hi Alexey, I do what you said and without -flag, the error "Unable to init server: Could not connect: Connection refused
(chart_yolov4-obj.png:2416): Gtk-WARNING **: 09:45:31.916: cannot open display: " show up. Do you have any idea what happen?

**Comment by jojo0513** [comment]
> Did you try to search for that error?
> 
> https://github.com/AlexeyAB/darknet/issues?q=cannot+open+display

I did, then someone suggested that I conclude the code with "-dont show." However, after adding it, the code continues to produce the same error and no CAM plot appears...

**Comment by stephanecharette** [comment]
You have to spell it correctly:  `-dont_show`.

**Comment by stephanecharette** [comment]
Or you have to use X-forwarding when you SSH into the device.  Or you have to run it on a device that has X installed.

**Comment by jojo0513** [comment]
> 

Yes, in my code, it is `-dont show`. I made a mistake in my last response. Additionally, I use Google colab; do I still need to instal X?

**Comment by stephanecharette** [comment]
You still typed it wrong in your last comment, so I'm not convinced you're using it correctly.  It should be `-dont_show`.  No, you don't have to install X if you use `-dont_show`.
